controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -162,98 +167,72 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
-                                    "Transactions on different epoch are being committed");
-                        }
-                    }
-                });
-    }
+    private CompletableFuture<VersionedMetadata<CommittingTransactionsRecord>> rollTransactions(String scope, String stream, HistoryRecord txnEpoch, HistoryRecord activeEpoch,
+                                                                                            VersionedMetadata<CommittingTransactionsRecord> existing, OperationContext context) {
+        CompletableFuture<VersionedMetadata<CommittingTransactionsRecord>> future = CompletableFuture.completedFuture(existing);
+        if (!existing.getObject().isRollingTransactions()) {
+            future = future.thenCompose(
+                    x -> streamMetadataStore.startRollingTxn(scope, stream, activeEpoch.getEpoch(),
+                            existing, context, executor));
+        }
 
-    /**
-     * 1. check idempotence --> check if all transactions are already committed, then rolling txn has already happened.
-     * 2. create duplicate txn epoch segments in segment store and commit transactions into those segments.
-     * 3. create duplicate active epoch segments
-     * 4. update history table with one complete and one partial epoch
-     * 5. seal active segments
-     * 6. complete partial record in history table
-     */
-    private CompletableFuture<Void> rollTransactions(String scope, String stream, HistoryRecord txnEpoch, HistoryRecord activeEpoch,
-                                                     List<UUID> transactionsToCommit, OperationContext context) {
-        // Idempotent: check if all transactions are already committed. if so, do nothing and return immediately
-        // Just checking the last transaction id in list of transactions to commit suffices as we perform processing of
-        // transactions commit in same order as in the last. So if last transaction is already committed, then everything
-        // preceeding it is also committed.
-        UUID lastTransactionId = transactionsToCommit.get(transactionsToCommit.size() - 1);
-        return streamMetadataStore.transactionStatus(scope, stream, lastTransactionId, context, executor)
-                .thenCompose(status -> {
-                    if (status.equals(TxnStatus.COMMITTING)) {
-                        return runRollingTxn(scope, stream, txnEpoch, activeEpoch, transactionsToCommit, context);
-                    } else {
-                        return CompletableFuture.completedFuture(null);
-                    }
-                });
+        return future.thenCompose(record -> {
+            if (activeEpoch.getEpoch() > record.getObject().getActiveEpoch()) {
+                return CompletableFuture.completedFuture(record);
+            } else {
+                return runRollingTxn(scope, stream, txnEpoch, activeEpoch, record, context)
```

Comment:
We want to run RollingTxn even if it was started in a previous iteration of this processor. 
So we want to call start only if it was not started.. and then have the block for calling runRollingTxn.

Following is the code structure:
```
CompletableFuture<CommittingTxnRecord> future; 
if (rollingTxn not started) { 
  // note: rollingTxn may have been started in a previous iteration and this may be a rerun.. 
  // we only call startRollingTxn if it was not running already.  
future = startRollingTxn()
} 
if (committingTxnRecord.activeEpoch < activeEpoch) {
    // this rolling txn is already complete..  
    future = completableFuture.completedFuture()
} else {
    future = runRollingTxn();
}
```